### PR TITLE
[bugfix]: Fixes hot reloading

### LINKED ISSUE #1590: [bug]: Uncaught Invariant Violation: You must pass a component to the function returned by connect. Instead received undefined
**Describe the bug**

Occasionally, with a dev server running like `yarn run watch:falkowski` a code change forcing a hot-reload will trigger this error:

> Uncaught Invariant Violation: You must pass a component to the function returned by connect. Instead received undefined

Which traces into the `client.js` file (I think, will update if I see it again).

Hard refreshes don't fix it, but quitting and restarting the watcher works (every time). Everything is fine for a while, then it'll randomly happen later. Not sure why.

**To Reproduce**

Not sure exactly. Have a watcher running, and write code. It only happens during a hot-reload (not launching a server).

**Expected behavior**
Hot-reloading stays hot and reloading.

**Screenshots**

**Additional context**
Reference to Slack thread: https://magentocommeng.slack.com/archives/C71HNKYS2/p1566412835381800?thread_ts=1566411734.381300&cid=C71HNKYS2

For me, it occurred working on the `create-account` route.

In Slack, others mentioned it seems to occur on Root Components.

**Possible solutions**
Add any ideas about possible solutions to the problem here.

**Please complete the following device information:**
 - Magento Version: 2.3.2
 - PWA Studio Version: August 12 (am) 2019
 - NPM version `npm -v`: 6.9.0
 - Node Version `node -v`: 10.16.1

**Please let us know what packages this bug is in regards to:**
- [ ] `venia-concept`
- [ ] `venia-ui`
- [x] `pwa-buildpack`
- [ ] `peregrine`
- [ ] `pwa-devdocs`
- [ ] `upward-js`
- [ ] `upward-spec`

### PR DESCRIPTION
**Pull request by sirugh**
## Description

Before I begin let's reproduce this so you know what I'm solving here.

1) On `develop`, open your local dev server with `yarn watch:all` and navigate to the `/create-account` page.
2) Now open `createAccountPage.js` and add `import Button from '../Button';` to the imports and `<Button>Foo</Button>` just beneath the container `div`. 
3) When you save this file and the app should start reloading and you should see an error in console (`staticContext` warning is resolved in #1966) _AND_ the page should look completely broken:

[Image from Gyazo](https://gyazo.com/aa216c5e104f76bdb0aacb037424ec97)

4) Now, check out this PR (reset your local changes too) and re-run the steps. The page should render and you should see the button appear without a reload!

**So what went wrong?**

This was an elusive [hot reloading](https://webpack.js.org/api/hot-module-replacement/) bug that was difficult to reproduce. I was able to finally repro by making changes to direct child of `Route` that was being lazy loaded. I assumed it was something to do with hot reloading as you could restart the watcher and the page would render fine (until you made some changes). The odd thing was that you could get the `Button` component "shell" to display but the child components and any other components on that page such as the `CreateAccountForm` would just not render.

After poking around I found [this comment](https://github.com/gaearon/react-hot-loader/issues/1139#issuecomment-449897132) on an issue on the "React hot loader" plugin. The comment suggests that React does some "double buffering" where it displays/updates an _alternate_ before displaying the final state. This sounded a lot like what was happening (although I can't say 100% without diving into React internals), particularly because of the "partial" render state that was returned after making changes.

I also noticed that our hot reloading of React components was not actaully hot reloading. In fact it seemed to be forcing a refresh after each change instead of swapping out the modified component.

So, to address the above issues I did the following:

1) I added [react-hot-loader](https://github.com/gaearon/react-hot-loader) to the babel plugins for `development`. Now, changes to lazy load components should work!
2) I added some additional code necessary to actually hot reload our React components. Now you should no longer see warnings like `[HMR] Error: Aborted because ... is not accepted`.

Caveats:

* Although suggested by the `react-hot-loader` readme, I did not use the hook support loader as ours seem to work without the addition. You can demonstrate that by adding some simple hook like `useEffect(() => console.log('yo'), ['hot']);` to a component. It should log to console without a hard reload.

## Related Issue

Closes #1590.

### Verification Stakeholders
@brendanfalkowski @zetlen 

### Verification Steps

After you've got the PR checked out, make some changes (independently) to the following to see the reloading work:
* `createAccountPage.js` - Add a `Button`. Represents a change to a visual component.
* `useCreateAccountPage.js` - Add an effect. Represents a change within a talon/hook.
* `reducers/app.js` - Set `drawer: 'cart'`. Represents a change within the reducer/store.

For each of the above you should see the app perform a swap rather than a full refresh.

## Screenshots / Screen Captures (if appropriate)
Shown above.

## Checklist

* I have updated the documentation accordingly, if necessary.
* I have added tests to cover my changes, if necessary.

### CODE REVIEW

File: packages/venia-concept/src/index.js [hunk 1]
```diff
@@ -45,14 +45,23 @@ const apolloLink = ApolloLink.from([
...
+            <AppContextProvider>
+                <Component />
+            </AppContextProvider>
+        </Adapter>,
+        document.getElementById('root')
+    );
+
+renderApp(App);
+if (module.hot) {
+    console.log('HOT');
```
Comment (sirugh) [inline packages/venia-concept/src/index.js]: 🔥 
```suggestion
```

File: packages/venia-concept/src/index.js [hunk 2]
```diff
@@ -45,14 +45,22 @@ const apolloLink = ApolloLink.from([
...
+            <AppContextProvider>
+                <Component />
+            </AppContextProvider>
+        </Adapter>,
+        document.getElementById('root')
+    );
+
+renderApp(App);
+if (module.hot) {
+    module.hot.accept('@magento/venia-ui/lib/components/App/container', () =>
```
Comment (sirugh) [inline packages/venia-concept/src/index.js]: I need to double check that this is the correct usage.

File: packages/venia-concept/package.json [hunk 3]
```diff
@@ -96,6 +96,7 @@
     "react": "~16.9.0",
     "react-dom": "~16.9.0",
     "react-feather": "~2.0.3",
+    "react-hot-loader": "~4.12.17",
```
Comment (sirugh) [inline packages/venia-concept/package.json]: Did I add this dependency to all the right places?

File: packages/venia-concept/src/index.js (hunk 2 above)
Comment (sirugh) [reply packages/venia-concept/src/index.js]: `module.hot.accept()` may also work but I'm concerned about the other side effects such as complete reloading of the graphql store.

Comment (sirugh) [reply packages/venia-concept/src/index.js]: Turns out this is fine, and even necessary, if we want changes to store to trigger hot reload.

File: packages/babel-preset-peregrine/lib/index.js [hunk 4]
```diff
@@ -35,7 +35,7 @@ const config = (api, opts = {}) => {
 
     const envConfigs = {
         development: {
-            plugins,
+            plugins: [...plugins, 'react-hot-loader/babel'],
```
Comment (sirugh) [inline packages/babel-preset-peregrine/lib/index.js]: _This_ fixes the breaking issue that forced us to restart the watcher.

File: packages/venia-concept/src/index.js [hunk 5]
```diff
@@ -62,3 +62,8 @@ window.addEventListener('online', () => {
 window.addEventListener('offline', () => {
     store.dispatch(app.setOffline());
 });
+
+if (module.hot) {
+    // When any of the dependencies to this entry file change we should hot reload.
+    module.hot.accept();
```
Comment (sirugh) [inline packages/venia-concept/src/index.js]: This simply makes it so the page doesn't hard refresh on every change.

Comment (tjwiebell) [reply packages/venia-concept/src/index.js]: Just wanted to document from our chat in the office; this code will not make it into the production bundle, given module.hot will not be loaded.

**Pull request merged**